sql/catalog/desc(s|riptortree): add data structure, refactor collection internals

ajwerner opened the pull request:
This is a step on the way to a cleaner and clearer future in the collection. I haven't finished splitting out all of the pieces but this is already getting bit. This makes it clearer what is going on.

ajwerner committed:
sql/catalog/descs: move some code into new files

Preliminary before more refactoring.

Release note: None

ajwerner committed:
sql/catalog/descriptortree: add descriptortree.Tree data structure

Also added ample tests. This thing is going to be handy.

Release note: None

ajwerner committed:
sql/catalog/descs: adopt descriptorTree for leased descriptors

Ends up being pretty good cleanup.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/66112)
<!-- Reviewable:end -->

ajwerner committed:
sql/catalog/descs: extract a file and concept for kvDescriptors

This allows us to isolate, a bit, the concept of reading from the
store and caching that read data.

Release note: None

ajwerner committed:
sql/catalog/descs: extract synthetic descriptor concept, adopt descriptortree

Release note: None

ajwerner committed:
sql/catalog/descs: refine leasedDescriptors to control actual leasing

Trying to untangle dependencies.

Release note: None

ajwerner committed:
sql/catalog/descs: break up monolithic collection.go

The file was so big and unorderly it was hard to see what was happening.
This commit is purely movement, no new code.

Release note: None

ajwerner commented:
This is really just preliminary refactoring. It should have zero behavior change. Some of the cleanup is inconsistent but it felt like a stopping point. There will be more. It's getting easier to see how to make further progress. 

Please take this one commit at a time.

fqazi reviewed (approved):
> Reviewed 4 of 4 files at r1, 10 of 10 files at r2, 7 of 7 files at r3, 3 of 3 files at r4, 3 of 3 files at r5, 2 of 2 files at r6, 9 of 9 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66112)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)



<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
TFTR!

bors r=fqazi

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66112#-:-Mbl57KY4MEfXYvernib:bo7owc4)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)



<!-- Sent from Reviewable.io -->
